### ISSUE TITLE
concat_on_disk should use int64 indptrs by default so that it can concatenate large files

### ISSUE THREAD

**Issue by agdenadel** [scverse/anndata#1470]
### Please make sure these conditions are met

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the latest version of anndata.
- [ ] (optional) I have confirmed this bug exists on the master branch of anndata.

### Report

Currently, when concatenating multiple anndata objects that collectively have more than int32.max non-zero values in their concatenated sparse matrix, an overflow exception occurs (in older versions of anndata, the overflow silently happens and leaves large swaths of the matrix as zeroes). 

anndata should detect this on the fly and cast indptrs to int64, or just use int64 indptrs for all sparse matrices.

Discussed in #1348.

A workaround is loading each h5ad file individually, casting the indptrs to int64, and saving new h5ad files before merging.
```python
def convert_h5ad_sparse_matrix_indices_to_int64(h5ad_file, new_file):
    adata = ad.read_h5ad(h5ad_file)
    adata.X.indptr = np.array(adata.X.indptr, dtype=np.int64)
    adata.write_h5ad(new_file)
```

### Versions

```python
anndata             0.10.6
```

**Comment by ivirshup** [comment]
I think we should definitely have `concat_on_disk` do this (though it's possible that could even just count how big everything is before writing anything?).

I don't think we need to always write `int64` though. Modifying existing files is a smaller use case, where we now error.

@ilan-gold would you be up for tackling this?

**Comment by agdenadel** [comment]
Following up on this issue, the workaround I proposed above doesn't work out of the box when trying to slice datasets. In particular, if `adata` is an anndata object with int64 indptrs, slicing doesn't work

`adata = adata[:, adata.var.highly_variable]`

`adata.X` yields

`ValueError: Output dtype not compatible with inputs.`

**Comment by agdenadel** [comment]
@ivirshup I just found your issue in scipy

https://github.com/scipy/scipy/issues/20182

I think upgrading to scipy  `1.13.0` solved the subsetting issue

**Comment by minigel** [comment]
@ilan-gold Thank you for fix #1493 as I am having this same issue. I tried concatenating some large files with `concat_on_disk` using the `merge.py` in this fix, but encountered the following error:
 

```
Session information updated at 2024-05-09 01:16
Traceback (most recent call last):
  File "/home/user/miniconda3/envs/scverse/lib/python3.12/site-packages/anndata/experimental/merge.py", line 632, in concat_on_disk
    _write_concat_arrays(
  File "/home/user/miniconda3/envs/scverse/lib/python3.12/site-packages/anndata/experimental/merge.py", line 300, in _write_concat_arrays
    write_concat_sparse(
  File "/home/user/miniconda3/envs/scverse/lib/python3.12/site-packages/anndata/experimental/merge.py", line 235, in write_concat_sparse
    out_dataset.append(temp_elem)
  File "/home/user/miniconda3/envs/scverse/lib/python3.12/site-packages/anndata/_core/sparse_dataset.py", line 531, in append
    indptr.resize((orig_data_size + sparse_matrix.indptr.shape[0] - 1,))
  File "/home/user/miniconda3/envs/scverse/lib/python3.12/site-packages/h5py/_hl/dataset.py", line 666, in resize
    raise TypeError("Only chunked datasets can be resized")
TypeError: Only chunked datasets can be resized 
```
